staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/promise.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package promise
+
+import (
+	"fmt"
+	"sync/atomic"
+
+	"k8s.io/apiserver/pkg/util/flowcontrol/counter"
+)
+
+// promise implements the promise.WriteOnce interface.
+// This implementation is based on a condition variable.
+// This implementation tracks active goroutines:
+// the given counter is decremented for a goroutine waiting for this
+// varible to be set and incremented when such a goroutine is
+// unblocked.
+//
+// FIXME: Move this to testing/ package.
+type promise struct {
+	doneCh        chan struct{}
+	activeCounter counter.GoRoutineCounter // counter of active goroutines
+	waitingCount  int32                    // number of goroutines idle due to this being unset
+
+	// Should be access via atomic.
+	isSet         int32
+	value         interface{}
+}
+
+var _ WriteOnce = &promise{}
+
+// NewCountingWriteOnce makes a new promise.LockingWriteOnce
+func NewCountingWriteOnce(activeCounter counter.GoRoutineCounter) WriteOnce {
+	return &promise{
+		doneCh:       make(chan struct{}),
+		activeCounter: activeCounter,
+	}
+}
+
+// FIXME:
+var timeoutErr = fmt.Errorf("timeout")
+
+func (p *promise) WaitAndGet(stopCh <-chan struct{}) (interface{}, error) {
+	if atomic.LoadInt32(&p.isSet) == 0 {
+		atomic.AddInt32(&p.waitingCount, 1)
+		p.activeCounter.Add(-1)
+
+		// There is a possible race that between checking p.isSet and increasing
+		// p.waitingCount p.Set was called and did the waitingCount decrease.
+		// To protect against this, we need to check if we shouldn't decrease here too.
+		defer p.unblockWaiting()
+
+		select {
+		case <-p.doneCh:
+			return p.value, nil
+		case <-stopCh:
+			return nil, timeoutErr
+		}
+	}
+	return p.value, nil
+}
+
+func (p *promise) IsSet() bool {
+	return atomic.LoadInt32(&p.isSet) != 0
+}
+
+func (p *promise) Set(value interface{}) bool {
+	if !atomic.CompareAndSwapInt32(&p.isSet, 0, 1) {
+		return false
+	}
+	p.value = value
+	close(p.doneCh)
+
+	p.unblockWaiting()
```

Comment:
The goroutine counter has to be updated before the waiting goroutines are unblocked (in this case by the `close(p.doneCh)`); doing it later creates timing splinters.